<body style=max-width:80ch;margin:auto>
<h1 class="entry-title" style="margin-bottom:15px;">C++/WinRT doesn’t let your coroutine cheat death, but it does get to say good-bye</h1>  <!-- .entry-meta -->

<p>I noted in a footnote that <a href="/code/how-to-get-your-c-winrt-asynchronous-operations-to-respond-more-quickly-to-cancellation-part-1"> it’s important to use RAII types for all of your cleanup in a coroutine</a> because cancellation will cause the coroutine to abruptly stop executing at the next suspension point. C++/WinRT doesn’t let your coroutine cheat death, but it does get to say good-bye.</p>
<p>C++/WinRT reports cancellation to the coroutine by throwing an <code>hresult_canceled</code> exception in the context of the coroutine itself. This takes advantage of the existing coroutine machinery to stow exceptions in the asynchronous operation so they can be observed by the code that is awaiting the result.</p>
<p>But it also means that you can see the exception fly past as you proceed inexorably to your death.</p>
<pre>IAsyncOperation&lt;int&gt; DoSomethingAsync()
{
    try {
        /* do stuff */
        co_return 42;
    } catch (hresult_canceled const&amp;) {
        printf("Pardonnez-moi, monsieur, je ne l'ai pas fait exprès.");
        throw;
    }
}
</pre>
<p>You can catch the <code>hresult_<wbr/>canceled</code> exception and perhaps do some final cleanup. When you’re done, you rethrow the exception to allow normal cleanup of the coroutine to proceed.</p>
<p>Of course, if your intention is to perform actual cleanup (rather than merely making a statement for posterity),¹ then you’ll need to catch <i>all</i> exceptions, not just cancellation.</p>
<pre>IAsyncOperation&lt;int&gt; DoSomethingAsync()
{
    ClaimWidget();
    try {
        /* do stuff */
        co_return 42;
    } catch (...) {
        DisclaimWidget();
        throw;
    }
}
</pre>
<p>At this point, you may as well just use an RAII type.</p>
<pre>struct ScopedWidgetClaim
{
    ScopedWidgetClaim() { ClaimWidget(); }
    ~ScopedWidgetClaim() { DisclaimWidget(); }

    // non-copyable, non-assignable
    ScopedWidgetClaim(ScopedWidgetClaim const&amp;) = delete;
    void operator=(ScopedWidgetClaim const&amp;) = delete;
};

IAsyncOperation&lt;int&gt; DoSomethingAsync()
{
    ScopedWidgetClaim claim;
    /* do stuff */
    co_return 42;
}
</pre>
<p>Maybe you think you can cheat death by swallowing the exception instead of rethrowing it?</p>
<pre>IAsyncOperation&lt;int&gt; DoSomethingAsync()
{
    try {
        /* do stuff */
        co_return 42;
    } catch (hresult_canceled const&amp;) {
        co_return 0;
    }
}
</pre>
<p>What happens here depends on where the cancellation exception was generated. If it was the result of a dependent coroutine failing with cancellation, then you successfully caught the failed dependent coroutine and can recover:</p>
<pre>IAsyncOperation&lt;int&gt; GetWidgetCountAsync()
{
    try {
        auto widgets = co_await GetAllWidgetsAsync();
        co_return widgets.Count();
    } catch (hresult_canceled const&amp;) {
        co_return 0;
    }
}
</pre>
<p>If the <code>Get­All­Widgets­Async</code> operation was cancelled (say, because it prompted the user for permission, and the user hit <i>Cancel</i>), then the <code>co_await GetAllWidgetsAsync()</code> will throw the cancellation exception, and you caught it and dealt with the problem by returning a widget count of zero.</p>
<p>On the other hand, if the cancellation exception was generated because somebody cancelled <i>you</i>, then your coroutine has already been put into the cancelled state. It’s too late to <code>co_return</code> a value; the coroutine has already been cancelled. You can still do the <code>co_return</code>, but the value you return won’t be the result of the coroutine.</p>
<p><b>Bonus chatter</b>: C++/WinRT does not provide a way to rescue a coroutine from cancellation. You might think of adding a feature like <i>co_await uncancel_current_coroutine()</i> so that a coroutine could treat cancellation as a way to abandon an operation and generate partial results, but C++/WinRT doesn’t do that. Once it’s cancelled, it’s cancelled.</p>
<p>In theory, it could be possible to add “uncancellation”, but it would come at the expense of responsiveness: Right now, when an <code>IAsyncXxx</code> is cancelled, the operation immediately transitions to the cancelled state. The completion handler is called immediately, allowing the code that is awaiting the result of the coroutine to proceed without delay. If operations could be uncancelled, then the transition to the cancelled state would have to be delayed until the coroutine itself decided whether to accept the cancellation or to uncancel it.</p>
<p>C++/WinRT opted not to slow down the common case in order to add a feature that would rarely be used.</p>
<p>¹ The technical term for “making a statement for posterity” is <i>logging</i>.</p>


</body>